File: Data Structures & Algorithms/course-schedule-ii/submission-5.cpp
class Solution {
public:
    unordered_map<int,set<int>> D ;
    void dfs(int key , vector<bool>& visited , vector<int>& res){
        if(visited[key]) return ;
        visited[key]=true;
        for(const int x : D[key] ){
            if(D.count(x) && !visited[x] && find(res.begin(), res.end(), x) == res.end()){
                dfs(x,visited,res);
                if(find(res.begin(), res.end(), x) == res.end()) return ;
            }
            else if(!D.count(x) && find(res.begin(), res.end(), x) == res.end() ) res.push_back(x);
            else if(find(res.begin(), res.end(), x) != res.end()) continue;
            else return ;
        }
        res.push_back(key);
    }
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        vector<bool> visited(numCourses,false);
        vector<int> res;
        for(const auto& p :prerequisites ){
            D[p[0]].insert(p[1]);
        }
        for(auto key : D){
            dfs(key.first,visited,res);
            if(find(res.begin(), res.end(), key.first) == res.end())
               return {};
        }
        for(int i=0;i<numCourses;i++){
            if(find(res.begin(), res.end(), i) == res.end())
                res.push_back(i);
        }
        return res ;
    }
};

File: Data Structures & Algorithms/course-schedule-iv/submission-5.py
class Solution:
    def checkIfPrerequisite(self, numCourses: int, prerequisites: List[List[int]], queries: List[List[int]]) -> List[bool]:
        mp={}
        keys=set()
        for p in prerequisites :
            if p[0] not in mp :
                mp[p[0]]={p[1]}
            else:
                mp[p[0]].add(p[1])
            keys.add(p[1])
            if p[1] in mp :
                mp[p[0]].update(mp[p[1]])
            if p[0] in keys:
                for key in mp :
                    if p[0] in mp[key] :
                        mp[key].update(mp[p[0]])

        res=[]
        for q in queries :
            if q[0] not in mp:
                res.append(False)
            elif q[1] in mp[q[0]] : 
                res.append(True)
            else :
                res.append(False)
        return res